---
layout: page
title: WGS
permalink: /exercieses/wgs
parent: Exercises
has_toc: true
nav_order: 4
---

1. TOC
{:toc}

---
## Genome assembly with 2nd and 3rd WGS data

In this session, we aim to assemble a bacterial genome using 2nd and 3rd whole genome sequencing (WGS) data. 
We will use it this as example to explore the WGS data analysis, and look into the difference between sequencing technologies.

## Software installation with `mamba` and data acquisition

Here will make 

## Exporation with a demo data

### Genome assembly with Illumina reads

#### Adapter removal with `trimmomatic`

#### Quality control with `bbmap`

#### Genome assembly with `spades`  

### Genome assembly with Nanopore reads

#### Optional: Adapter removal with `guppy` or `porechop`

#### Quality control with `seqkit`

#### Genome assembly with `flye`

#### Genome polishing with `racon` and `medaka`

#### Illumina reads polishing with `pilon`

#### Optional: hybrid assembly with `unicycler`

### Quality assessment of assembled genomes with `quast`

### Optional: Genome annotation with `prokka`

### Optional: Frameshift correction with `proovframe`



